File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/splash/SplashActivity.java
package openfoodfacts.github.scrachx.openfood.views.splash;

import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;

import com.google.android.material.snackbar.Snackbar;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.databinding.ActivitySplashBinding;
import openfoodfacts.github.scrachx.openfood.views.BaseActivity;
import openfoodfacts.github.scrachx.openfood.views.welcome.WelcomeActivity;
import pl.aprilapps.easyphotopicker.EasyImage;

import static com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG;

public class SplashActivity extends BaseActivity implements ISplashActivity.View {
    private ActivitySplashBinding binding;
    private String[] taglines;
    /*
    To show different slogans below the logo while content is being downloaded.
     */
    private final Runnable changeTagline = new Runnable() {
        int i = 0;

        @Override
        public void run() {
            i++;
            if (i > taglines.length - 1) {
                i = 0;
            }
            if (binding != null) {
                binding.tagline.setText(taglines[i]);
                binding.tagline.postDelayed(changeTagline, 1500);
            }
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySplashBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        if (getResources().getBoolean(R.bool.portrait_only)) {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }

        taglines = getResources().getStringArray(R.array.taglines_array);
        binding.tagline.post(changeTagline);

        ISplashActivity.Controller presenter = new SplashController(getSharedPreferences("prefs", 0), this, this);
        presenter.refreshData();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
    }

    @Override
    public void navigateToMainActivity() {
        EasyImage.configuration(this)
            .setImagesFolderName("OFF_Images")
            .saveInAppExternalFilesDir()
            .setCopyExistingPicturesToPublicLocation(true);
        WelcomeActivity.start(this);
        finish();
    }

    @Override
    public void showLoading() {
    }

    @Override
    public void hideLoading(boolean isError) {
        if (isError) {
            new Handler(Looper.getMainLooper()).post(() -> Snackbar.make(binding.getRoot(), R.string.errorWeb, LENGTH_LONG).show());
        }
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/splash/SplashActivity.kt
package openfoodfacts.github.scrachx.openfood.features.splash

import android.annotation.SuppressLint
import android.content.pm.ActivityInfo
import android.os.Bundle
import androidx.activity.viewModels
import androidx.core.view.isVisible
import androidx.lifecycle.lifecycleScope
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.databinding.ActivitySplashBinding
import openfoodfacts.github.scrachx.openfood.features.shared.BaseActivity
import openfoodfacts.github.scrachx.openfood.features.welcome.WelcomeActivity
import openfoodfacts.github.scrachx.openfood.utils.getAppPreferences
import pl.aprilapps.easyphotopicker.EasyImage
import kotlin.time.ExperimentalTime

@AndroidEntryPoint
class SplashActivity : BaseActivity() {
    private lateinit var binding: ActivitySplashBinding
    private val viewModel: SplashViewModel by viewModels()

    private val controller by lazy {
        SplashController(getAppPreferences(), this, this)
    }

    @SuppressLint("SourceLockedOrientationActivity")
    @ExperimentalTime
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySplashBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if (resources.getBoolean(R.bool.portrait_only)) {
            requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        }

        viewModel.setupTagLines(resources.getStringArray(R.array.taglines_array))
        viewModel.tagLine.observe(this, binding.tagline::setText)

        lifecycleScope.launch { controller.refreshData() }
    }

    suspend fun navigateToMainActivity() = withContext(Dispatchers.Main) {
        EasyImage.configuration(this@SplashActivity).apply {
            setImagesFolderName("OFF_Images")
            saveInAppExternalFilesDir()
            setCopyExistingPicturesToPublicLocation(true)
        }
        WelcomeActivity.start(this@SplashActivity)
        finish()
    }

    suspend fun showLoading() = withContext(Dispatchers.Main) {
        binding.loadingTxt.isVisible = true
        binding.loadingProgress.isVisible = true
    }

    suspend fun hideLoading(isError: Boolean) = withContext(Dispatchers.Main) {
        binding.loadingTxt.isVisible = false
        binding.loadingProgress.isVisible = false
        if (isError) {
            Snackbar.make(
                binding.root,
                R.string.errorWeb,
                BaseTransientBottomBar.LENGTH_LONG
            ).show()
        }
    }
}
